--- a/grub-core/kern/main.c	2012-06-28 10:17:43.000000000 +0200
+++ b/grub-core/kern/main.c	2012-09-06 00:43:10.499239061 +0200
@@ -212,7 +212,7 @@
 
   /* Hello.  */
   grub_setcolorstate (GRUB_TERM_COLOR_HIGHLIGHT);
-  grub_printf ("Welcome to GRUB!\n\n");
+  //grub_printf ("Welcome to GRUB!\n\n");
   grub_setcolorstate (GRUB_TERM_COLOR_STANDARD);
 
   /* Load pre-loaded modules and free the space.  */
diff -Naur a/grub-core/loader/i386/linux.c b/grub-core/loader/i386/linux.c
--- a/grub-core/loader/i386/linux.c	2012-06-28 10:17:43.000000000 +0200
+++ b/grub-core/loader/i386/linux.c	2012-09-06 00:44:52.610521282 +0200
@@ -906,20 +906,20 @@
 	  case 0:
 	  case GRUB_LINUX_VID_MODE_NORMAL:
 	    grub_env_set ("gfxpayload", "text");
-	    grub_printf_ (N_("%s is deprecated. "
+	    /*grub_printf_ (N_("%s is deprecated. "
 			     "Use set gfxpayload=%s before "
 			     "linux command instead.\n"),
-			  argv[i], "text");
+			  argv[i], "text");*/
 	    break;
 
 	  case 1:
 	  case GRUB_LINUX_VID_MODE_EXTENDED:
 	    /* FIXME: support 80x50 text. */
 	    grub_env_set ("gfxpayload", "text");
-	    grub_printf_ (N_("%s is deprecated. "
+	    /*grub_printf_ (N_("%s is deprecated. "
 			     "Use set gfxpayload=%s before "
 			     "linux command instead.\n"),
-			  argv[i], "text");
+			  argv[i], "text");*/
 	    break;
 	  default:
 	    /* Ignore invalid values.  */
@@ -929,10 +929,10 @@
 		grub_env_set ("gfxpayload", "text");
 		/* TRANSLATORS: "x" has to be entered in, like an identifier,
 		   so please don't use better Unicode codepoints.  */
-		grub_printf_ (N_("%s is deprecated. VGA mode %d isn't recognized. "
+		/*grub_printf_ (N_("%s is deprecated. VGA mode %d isn't recognized. "
 				 "Use set gfxpayload=WIDTHxHEIGHT[xDEPTH] "
 				 "before linux command instead.\n"),
-			     argv[i], vid_mode);
+			     argv[i], vid_mode);*/
 		break;
 	      }
 
@@ -946,10 +946,10 @@
 	    if (! buf)
 	      goto fail;
 
-	    grub_printf_ (N_("%s is deprecated. "
+	    /*grub_printf_ (N_("%s is deprecated. "
 			     "Use set gfxpayload=%s before "
 			     "linux command instead.\n"),
-			 argv[i], buf);
+			 argv[i], buf);*/
 	    err = grub_env_set ("gfxpayload", buf);
 	    grub_free (buf);
 	    if (err)
diff -Naur a/grub-core/normal/menu.c b/grub-core/normal/menu.c
--- a/grub-core/normal/menu.c	2012-06-25 15:53:58.000000000 +0200
+++ b/grub-core/normal/menu.c	2012-09-06 00:45:15.195362637 +0200
@@ -287,7 +287,7 @@
 {
   int fallback_entry;
 
-  callback->notify_booting (entry, callback_data);
+  //callback->notify_booting (entry, callback_data);
 
   grub_menu_execute_entry (entry, 1);
 
