diff --git a/grub/util/grub.d/10_linux.in b/grub/util/grub.d/10_linux.in
index de9044c..6a96222 100644
--- a/grub/util/grub.d/10_linux.in
+++ b/grub/util/grub.d/10_linux.in
@@ -47,7 +47,7 @@ esac
 # and mounting btrfs requires user space scanning, so force UUID in this case.
 if [ "x${GRUB_DEVICE_UUID}" = "x" ] || [ "x${GRUB_DISABLE_LINUX_UUID}" = "xtrue" ] \
     || ! blkid -U "${GRUB_DEVICE_UUID}" > /dev/null \
-    || ( test -e "${GRUB_DEVICE}" && uses_abstraction "${GRUB_DEVICE}" lvm ); then
+    || ( test -e "${GRUB_DEVICE}" && uses_abstraction "${GRUB_DEVICE}" lvm && ! uses_abstraction "${GRUB_DEVICE}" luks ); then
   LINUX_ROOT_DEVICE=${GRUB_DEVICE}
 else
   LINUX_ROOT_DEVICE=UUID=${GRUB_DEVICE_UUID}
